### DISCUSSION TITLE
Recommended way for out-of-process indexing

### DISCUSSION THREAD

**Discussion by faustbrian** [tendermint/tendermint#9068] [Q&A] [answered]
Tendermint out-of-the-box comes with an indexer for blocks and transactions but I was wondering if there's any recommended methods or things to keep in mind for out-of-process indexing if you don't want to use any of the stock indexers.

- For example how should blocks be indexed if you want to do it manually. Should the `header` from `BeginBlock` be stored as soon as it is received to replicate what Tendermint stores or should it only be stored after `Commit` has been called?
- Similarly for transactions Tendermint stores them associated with a block ID. Should this block ID be derived from the `BeginBlock` header or somewhere else?
- I assume both of the above questions would be obsolete with the ABCI++ where you could perform all those tasks in `FinalizeBlock` without having to store some transient state like the block header?

**Comment by creachadair** [comment, marked as answer]
In short, there's not an obvious "right" way to do what you're asking at the moment. There are a few approaches you might consider:

The built-in indexers work by [observing NewBlockHeader and Tx events](https://github.com/tendermint/tendermint/blob/master/internal/state/indexer/indexer_service.go#L133). The ABCI block/tx metadata are not written to the event sink (i.e., the indexer) until all the transactions for a block have been received. At that point, all the ABCI metadata from BeginBlock, EndBlock, and the transactions contained within the block are recorded to storage, if indexing is enabled. If you wanted to emulate the built-in indexer, you'd have to more or less reproduce that logic. That is feasible, though probably a lot of work.

There's been some ongoing interest in making the indexing interface more pluggable (cf. #7135, [RFC 012](https://github.com/tendermint/tendermint/blob/master/docs/rfc/rfc-012-custom-indexing.md)). Right now, plugging in a new indexer requires rebuilding Tendermint itself (since the indexers are compiled-in). In the future, another option would be to take advantage of the [ADR 075 events interface](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-075-rpc-subscription.md) to allow a separate external process to process events. That interface is implemented in the candidate branch for v0.36 but not yet deployed in production.

Another option in the near term might be to use the [psql event sink](https://github.com/tendermint/tendermint/blob/master/internal/state/indexer/sink/psql/psql.go#L38), which stores event metadata in a PostgreSQL database. That would allow you to write SQL queries against the stored metadata, which could be simpler to wrangle than building a custom indexer on the application side.  This is arguably more of a workaround for the use case you're describing, but it might be easier to test out.
